### PR TITLE
Only R2R public methods, methods that override public methods, and internal methods that aren't always inlined

### PR DESCRIPTION
**Pull request by jkoritzinsky**
Today, we generate bodies for all methods in crossgen2. This results in wasted space in libraries formethods that are always inlined. For System.Private.CoreLib on Windows x64 (release build), this amounts to 4653 methods totaling 298kb of size-on-disk space that will never be used.

This PR changes the rooting rules for crossgen2 as follows:

- If a method is exposed outside the assembly as public, protected in a public type, or internal in an assembly with an InternalsVisibleToAttribute applied, it will be rooted.
- If a method implements an interface method or overrides a virtual method that meets the criteria above, it will also be rooted.
- If a method is explicitly mentioned in an embedded linker descriptor file, it will be rooted as well. This option ensures that methods that the runtime calls in CoreLib are pre-compiled in addition to being preserved by the linker. This may also make other methods continue to be precompiled that shouldn't be, but I think the number of methods in that case aren't particularly high, especially given that we were precompiling everything beforehand.

I've attached [a diff from R2R dump](https://github.com/dotnet/runtime/files/9637468/r2r-diff-main.txt) so it's easy to see the change in CoreLib.

<details>
<summary>Original Description</summary>
Today, we generate bodies for all P/Invokes that have blittable IL stubs in crossgen2. This results in wasted space in libraries for P/Invokes that are always inlined. For System.Private.CoreLib on Windows x64 (release build), this amounts to 374 methods totaling 40kb of size-on-disk space that will never be used (we only pre-generate 10 P/Invoke stubs after this change).

This PR utilizes the "recompile this method" feature to delay compilation of any P/Invoke's stubs until the P/Invoke is actually used in a way that causes it to not be inlined.

With this change, there is no change in the number of IL stubs emitted in a Hello World app on Windows x64.

Open Questions:

- Does the JIT notify the EE across the interface when inlining fails due to the block being cold, the JIT being told to optimize for size, or the JIT being told to generate debuggable code?
- Should we still emit code for any public (or internal w/ InternalsVisibleTo) blittable P/Invokes? In these cases, we can't guarantee that they won't be called by non R2R'd user code, so without updating this PR to account for this, these stubs will now be generated at runtime. This case doesn't apply for any of our product code as all of our P/Invokes are internal or private and we don't use InternalsVisibleTo in the dotnet/runtime repo.

I've included [a diff from R2R dump](https://github.com/dotnet/runtime/files/9590592/r2r-diff.txt) so it's easy to see the change this PR causes in CoreLib.

Contributes to https://github.com/dotnet/runtime/issues/69758 by reducing further the cases where we pre-generate stubs in crossgen2.
</details>

### CODE REVIEW

File: src/coreclr/tools/aot/ILCompiler.ReadyToRun/JitInterface/CorInfoImpl.ReadyToRun.cs
Comment (davidwrighton) [inline src/coreclr/tools/aot/ILCompiler.ReadyToRun/JitInterface/CorInfoImpl.ReadyToRun.cs:2936]: I don't believe this is the correct approach, we should use the standard dependencies mechanism for this, but it looks like we might not have the correct hooks for this at the moment. Contact me on Monday and I'll look into this with you.

**Review by davidwrighton — DISMISSED** [review]
Nice... but I think the recompile tech is the wrong hook to use here.

File: src/coreclr/tools/aot/ILCompiler.ReadyToRun/Interop/IL/MarshalHelpers.ReadyToRun.cs
Comment (jkotas) [inline src/coreclr/tools/aot/ILCompiler.ReadyToRun/Interop/IL/MarshalHelpers.ReadyToRun.cs:11]: Can this be moved and deduped with  `src\coreclr\tools\Common\TypeSystem\Interop\IL\Marshaller.cs`?

**Comment by jkotas** [comment]
I think a better way to do this would be to compile these stubs, but do not save them into the final image if we find that they are not referenced from anywhere. It would address the problem of JIT deciding to not inline the stub in cold code, etc.

Similar optimization can be applied to all inlineable method (the same concerns about public or InternalVisible apply). The blittable PInvoke stubs are really just a special-case of inlineable methods.

File: src/coreclr/tools/aot/ILCompiler.ReadyToRun/JitInterface/CorInfoImpl.ReadyToRun.cs
Comment (MichalStrehovsky) [reply src/coreclr/tools/aot/ILCompiler.ReadyToRun/JitInterface/CorInfoImpl.ReadyToRun.cs:2936]: We have `_additionalDependencies` in CorInfoImpl.cs that is used outside READYTORUN. It might be the thing to move out of `if READYTORUN` (and adding the respective `InitializeNonRelocationDependencies`) to solve this.

**Comment by davidwrighton** [comment]
@jkotas, the right way to do that would be to use the `_additionalDependencies` model. @jkoritzinsky, so the way to fix it is to enable the `_additionalDependencies` logic in the jit and `MethodWithGCInfo.cs` ... but we don't need the logic around `GetDependenciesDueToAccess` that is used to populate _additionalDependencies for NativeAOT.

**Comment by jkotas** [comment]
References to PInvoke stubs are relocs. Is there a reason for not using reloc dependencies for this? 

I agree that you can make it work using `_additionalDependencies`. I am just wondering whether reloc dependencies would be a better fit.

**Comment by MichalStrehovsky** [comment]
> References to PInvoke stubs are relocs. Is there a reason for not using reloc dependencies for this?

If these show up as relocations within the produced method code, it would definitely be worth investigating why it doesn't "just work" automatically (it should be requesting a method body for the p/invoke and none of the extra code should be needed to handle that).

**Comment by jkotas** [comment]
> it would definitely be worth investigating why it doesn't "just work" automatically

crossgen2 adds all methods as roots: https://github.com/dotnet/runtime/blob/57bfe474518ab5b7cfe6bf7424a79ce3af9d6657/src/coreclr/tools/aot/ILCompiler.ReadyToRun/Compiler/ReadyToRunLibraryRootProvider.cs#L79-L128 . We would need to stop doing this for the dependency analyzer to kick in.

**Comment by jkoritzinsky** [comment]
I've validated that the behavior is identical to what I had before if I just don't root P/Invokes, so I've removed the JitInterface changes.

**Comment by jkoritzinsky** [comment]
I did a prototype of only rooting methods that are visible outside of the assembly and found that we can save probably an additional 300kb from an additional 6k methods or so that are always inlined. This doesn't account for any rooting of the internal/private methods called by the runtime, so I'd need to port that infrastructure over from ILC (as we already support the .NET Linker rooting format there, and we already generate a file of that format as part of the CoreLib build for these exact methods) to get exact numbers.

I've attached a copy of the diff against this PR that I got in my prototype. I'm a little suspicious of some of the results, so I definitely need to do more work if we want to go in this direction (in addition to rooting method called from CoreLib).

[r2r-diff.txt](https://github.com/dotnet/runtime/files/9630182/r2r-diff.txt)

**Comment by jkoritzinsky** [comment]
I've updated this PR to be more generic and only root methods that may be exposed externally instead of only skipping P/Invokes. Can I get another review pass?

**Comment by jkotas** [comment]
Is this statement correct? "Reflection invoke on private methods will fall back to JIT unless the method is rooted in some other way (e.g. called from somewhere else)."

The diff you have shared is surprising. If I read it correctly, methods like `System.Diagnostics.Tracing.RuntimeEventSource.OnEventCommand` and `System.Buffers.TlsOverPerCoreLockedStacksArrayPool`1<__Canon>.Trim` are not compiled anymore. Why is that?

**Comment by jkoritzinsky** [comment]
> Is this statement correct? "Reflection invoke on private methods will fall back to JIT unless the method is rooted in some other way (e.g. called from somewhere else)."

Yes, that statement is correct. If we want to make that statement incorrect (ie track reflection), we could import more of the linker compatibility code from ILC to assist in reflection analysis.

> The diff you have shared is surprising. If I read it correctly, methods like `System.Diagnostics.Tracing.RuntimeEventSource.OnEventCommand` and `System.Buffers.TlsOverPerCoreLockedStacksArrayPool`1<__Canon>.Trim` are not compiled anymore. Why is that?

I need to figure that out. I'm seeing quite a few interesting methods in that list that all seem to match the following description:

- All callsites of the method are simple methods whose only statement is a call to this method.

Do you have any ideas why methods in this case wouldn't get dependency nodes added? Many of the methods (like `System.Buffers.TlsOverPerCoreLockedStacksArrayPool`1<__Canon>.Trim`) seem to be too large to be inlined.

**Comment by jkotas** [comment]
> Do you have any ideas why methods in this case wouldn't get dependency nodes added? 

It may a problem with tailcall optimization confusing the dependency tracker.

**Comment by jkoritzinsky** [comment]
That would make sense. I'll see if I can trace down what's happening there.

**Comment by jkoritzinsky** [comment]
I think the tailcalls might not use the reloc's mechanism and as a result are being missed. @davidwrighton mentioned some other dependency mechanism in ILC above that might be good for this. We may have to update the R2R side of the JIT interface after all to add these dependencies during code-generation. I'll give it a shot.

**Comment by jkotas** [comment]
The tailcalls targets are relocs as well. My guess is that the type of the reloc for tailcall is different from a regular call and not understood by the dependency tracker, but it is still a reloc. I do not think the extra reloc scheme is required to deal with tailcalls. Relocs should be enough.

File: src/coreclr/tools/Common/TypeSystem/Ecma/EcmaType.cs
Comment (MichalStrehovsky) [inline src/coreclr/tools/Common/TypeSystem/Ecma/EcmaType.cs:31]: We should inline this logic to where it's needed (or into an extension method). It doesn't look like we need to cache this and metadata reader + handle is available as a public API on `EcmaType`.

**Comment by MichalStrehovsky** [comment]
Cc @vitek-karas @tlakollo @dotnet/ilc-contrib - we now have a need to share code between IL Linker + NativeAOT + crossgen2.

**Comment by jkoritzinsky** [comment]
> The diff you have shared is surprising. If I read it correctly, methods like `System.Diagnostics.Tracing.RuntimeEventSource.OnEventCommand` and <code>System.Buffers.TlsOverPerCoreLockedStacksArrayPool`1<__Canon>.Trim</code> are not compiled anymore. Why is that?

I've figured out that I was being too aggressive on handling overrides (I didn't account for overrides with a `MethodImpl` record). My most recent commit preserves all virtual methods in addition to methods with explicit MethodImpl records (so static abstracts are still handled). This is more conservative, but it ensures that we don't accidentally drop any methods we might need. We can come back and make this more specific later. The change as of 59b4a64 is still a 151kb size savings.

[r2r-diff.txt](https://github.com/dotnet/runtime/files/9679007/r2r-diff.txt)

**Comment by jkotas** [comment]
Why does `void System.TimeSpan..ctor(int, int, int)` show up in the diff?

**Comment by jkoritzinsky** [comment]
I'm only seeing `System.TimeSpan..ctor(int, int, int)` in the second section where they're identical. I tried using the "skip same disasm" option to remove the diff entries for identical methods, but it was crashing when I tried it.

File: src/libraries/System.Private.Xml/src/System/Xml/Schema/XmlValueConverter.cs
Comment (jkotas) [inline src/libraries/System.Private.Xml/src/System/Xml/Schema/XmlValueConverter.cs:274]: This is complaining about array types. Are all possible array types that this can create really guaranteed to be used by something else?

Comment (jkoritzinsky) [reply src/libraries/System.Private.Xml/src/System/Xml/Schema/XmlValueConverter.cs:274]: I believe so, but I'll remove the suppression and propagate this up to callers to move the suppression closer to all places where the System.Type that's the element type is determined.

Comment (jkotas) [reply src/libraries/System.Private.Xml/src/System/Xml/Schema/XmlValueConverter.cs:274]: It may be best to submit the libraries fixes in a separate PR so that they get proper scrutiny.

Comment (jkoritzinsky) [reply src/libraries/System.Private.Xml/src/System/Xml/Schema/XmlValueConverter.cs:274]: I've revised the message here after some investigation. I'll split the commits into a different PR as well as pushing here to make sure there aren't any other issues.

File: src/coreclr/tools/aot/ILCompiler.ReadyToRun/Compiler/ReadyToRunMainMethodRootProvider.cs
Comment (MichalStrehovsky) [inline src/coreclr/tools/aot/ILCompiler.ReadyToRun/Compiler/ReadyToRunMainMethodRootProvider.cs:12]: Nit: might be better to just do it in ReadyToRunRootProvider, unless you can think of a scenario where we would want to do ReadyToRun roots and not the entrypoint or vice versa (I can't think of one).

File: src/coreclr/tools/aot/ILCompiler.ReadyToRun/Compiler/ReadyToRunXmlRootProvider.cs
Comment (MichalStrehovsky) [inline src/coreclr/tools/aot/ILCompiler.ReadyToRun/Compiler/ReadyToRunXmlRootProvider.cs:85]: Some of these appear unreferenced

(It's one of the reasons I'm not a fan of introducing `const` for what is a public API that will never change - introducing const is always a maintenance cost that should be weighed against the benefit - but this is the IL Linker codebase style...).

**Comment by MichalStrehovsky** [comment]
For the "File renamed without changes." changes, can you please check for copies under ReferenceSource and move them as well? These files are shared with illinker, the ReferenceSource is the literal copy from the linker repo - we keep them up to date by copying over the ReferenceSource files and manually applying the diffs.

E.g.
`tools\aot\ILCompiler.Compiler\Compiler\ReferenceSource\ProcessLinkerXmlBase.cs` and
`tools\aot\ILCompiler.Compiler\Compiler\ProcessLinkerXmlBase.cs`

Please also make a copy of README.md that has the commit hash from the illinker repo.

I'm kind of thinking whether we should move everything that is on the sharing plan to Common, even if it's not used by crossgen2 - so that we don't have files littered in yet another directory. Cc @dotnet/ilc-contrib @vitek-karas for thoughts.

**Review by MichalStrehovsky — APPROVED** [review]
Looks good to me otherwise, but I only skimmed the crossgen2-specific changes and it would be better if a crossgen dev has a look as well.

**Comment by jkoritzinsky** [comment]
cc: @dotnet/crossgen-contrib can I get a second review?

**Comment by jkotas** [comment]
Could you please share data about the size reductions for the netcoreapp with the latest change?

I am worried about the worst-case effects of the visibility-based rooting policy. For example, it is not unusual to see frameworks that use private reflection for binding. This will skip precompiling everything that these frameworks bind to (unless it is reachable by other means).

I think we may need explicit policy to control the visibility-based rooting policy and/or enable it only in limited fashion by default.

**Comment by jkoritzinsky** [comment]
The size reduction for R2Rd NETCoreApp as a whole is pretty small:

Test case is uncompressed `dotnet-runtime-internal` zip from a Release x64 build.
66.4MB - with change
67.1MB - without change

So about 0.7MB savings for the default configuration. 151KB of that is in System.Private.CoreLib.

If you feel it's important, I can make this a configurable option.

**Comment by jkotas** [comment]
It looks good enough to me. As @MichalStrehovsky said, it would be better if a crossgen dev has a look as well.

**Review by trylek — APPROVED** [review]
Looks great to me and nicely ties into my current experiments regarding reduction of R2R container size, thanks Jeremy!

**Pull request merged**
